### ISSUE TITLE
line breaks in command block

### ISSUE THREAD

**Issue by odoublewen** [dnanexus/dxWDL#274]
Say I have a command block like this:

```
  command <<<
  samtools index ${bam}
  bcftools mpileup -O u --annotate DP -f ${ref} -r ${region} ${bam} | bcftools call -O u --skip-variants indels --consensus-caller | bcftools filter -O z --exclude '%QUAL<25 | (FORMAT/DP)<10' -o ${prefix}.bcf
  bcftools index ${prefix}.bcf
  >>>
```

For readability, I would like to break up that second command across multiple lines, like:

```
  command <<<
  samtools index ${bam}
  bcftools mpileup -O u --annotate DP -f ${ref} -r ${region} ${bam} \
    | bcftools call -O u --skip-variants indels --consensus-caller \
    | bcftools filter -O z --exclude '%QUAL<25 | (FORMAT/DP)<10' -o ${prefix}.bcf
  bcftools index ${prefix}.bcf
  >>>
```

But it does not work.  Not sure if line breaks getting munged by the docker script or something?  Is there some escaping I can do to make sure the `\` characters are sent to the shell?

Are line breaks inside command blocks something that WDL spec does not support? Or is this a dxWDL issue?

I can provide a workable example of this if needed, but I'm thinking there is a simple explanation.  Thanks!

**Comment by orodeh** [comment]
@odoublewen, have you tried WDL version 1.0? I believe this problem was addressed there.

Thanks,
     Ohad.

**Comment by odoublewen** [comment]
> have you tried WDL version 1.0?

If by this you mean, have I tried using dxWDL >= 1.0 then yes, I'm using `dxWDL-1.11.jar` and I can confirm that the above command block works without the line breaks (first example), and does not work with the line breaks (second example.)

I'll try to make a silly reproducible example of the behavior I'm seeing.

**Comment by orodeh** [comment]
Sorry, I didn't mean the `dxWDL` version, I meant the WDL version. 

Please try placing the line `version 1.0` at the beginning of your WDL script. The parser is different for version 1.0, vs. draft-2. Several things work much better in version 1.0.

**Comment by odoublewen** [comment]
Okay, thanks,  I hadn't noticed that I had to add the `version 1.0` line use the updated spec.  (Seems like a poor namespace choice on WDL's part.... wish it was `wdl_version 1.0` or something like that so as not to get confused with the versioning of the workflow itself.)

Took me a while to realize that I cannot use `${var}` style expressions -- I have to switch to tilde (`~{var}`) when inside of  `<<<  >>>` blocks.  (per [WDL docs](https://github.com/openwdl/wdl/blob/master/versions/development/SPEC.md#expression-placeholders))

But, in the end, this did not fix the issue for me.  Still getting the same error message when breaking up the line as shown above.  That message is `mpileup: unrecognized option: skip-variants` -- suggesting that the line break and piping is not being interpreted correctly.  When I put the entire pipe expression on one line, it works fine.

**Comment by orodeh** [comment]
I tried the task `multi_line` with Cromwell 42, and the current dxWDL; it did work. Is this what you are trying to do?

```wdl
task multi_line {

    command {
    ls /home \
          -l \
          -r \
          -h
    }

    output {
        String results = read_string(stdout())
    }
}
```

**Comment by odoublewen** [comment]
My example has to do with the pipe character, and now I've discovered it only manifests when the pipe character is after the line break.  Here's my example:

```
version 1.0

workflow TestLineSplits {
  call EchoLineSplit
}

task EchoLineSplit {

  command {
  echo 1 hello world | sed 's/world/wdl/'
  echo 2 hello \
  world \
  | sed 's/world/wdl/'
  echo 3 hello \
  world | \
  sed 's/world/wdl/'
  }

  output {
    String results = read_string(stdout())
  }

  runtime {
    docker: "debian:stretch-slim"
    memory: "4 GB"
    cpu: "4"
    disks: "local-disk 10 SSD"
  }
}
```

And the output when I compile with dxWDL and run on DNAnexus is:

```
1 hello wdl
2 hello world sed s/world/wdl/
3 hello wdl
```

The second case, with the pipe character after the line break, fails.  It's like the pipe character is swallowed up.  It works perfectly well in any shell script.  I suspect there is some trimming of the whitespace inside the command block, and the `|` is being removed.

Fine with me if this is a Wont-Fix -- I have a workaround now.

**Comment by orodeh** [comment]
I reproduced the problem. Thanks for reporting. 

I thought the new Cromwell/WOM library + version 1.0 was going to make all of the special character problems go away. It now looks like there is something extra that needs to be done in dxWDL itself.

**Comment by orodeh** [comment]
Fixed in master.

**Comment by orodeh** [comment]
Released in version [1.13](https://github.com/dnanexus/dxWDL/releases/tag/1.13)
